enhancements/machine-api/out-of-tree-provider-support.md

```diff
@@ -0,0 +1,477 @@
+- Ensure that upgrade from a release running on the in-tree provider only to the out-of-tree succeeds, and the downgrade is also supported.
+- Each provider should have a working CI implementation, which will assist in testing the provisioning and upgrades with out-of-tree support enabled.
+
+### Graduation Criteria
+
+#### Tech preview
+
+Operator handles OpenStack installation in 4.7.
+
+#### Tech Preview -> GA
+
+The operator will handle basic tasks, such as create initial templates at bootstrap and running all of supported in-tree providers after installation.
+
+## FAQ
+
+Q: Can we disaster-recover a cluster without CCM running?
+A: We can't. Basically, without CCM running, nodes can only join the cluster, but they will be unschedulable. Which means it's impossible to start any workloads on the nodes. [Source](https://kubernetes.io/docs/tasks/administer-cluster/running-cloud-controller/#running-cloud-controller-manager)
+
+Q: What is non-functional in a cluster during bootstrapping until CCM is up?
+A: If CCM is not available, new nodes in the cluster will be left unschedulable.
+
+Q: Who talks to CCM?
+A: From an architectural point of view, only Kube API server is able to communicate with CCM. [Source](https://kubernetes.io/docs/concepts/architecture/cloud-controller/#design). All requests to CCM are sent by `kubelet`. KCM doesn't interact with CCM.
+
+Q: CCM relation to other openshift components, such as SDN and storage? How a non-operational CCM will affect cluster health, which components will take a hit?
+A: A couple of components rely on CCM at different time of the cluster lifetime:
+  - `kubelet` dependency requires removal of taints from newly created Nodes.
+  - CCMs manage `LoadBalancer` `Service` resources in all in-tree cloud implementations, except vSphere, so their creation will not be possible while the CCM is down. This will affect the [ingress-operator](https://github.com/openshift/cluster-ingress-operator). Cloud networking across `Nodes` is [disabled](https://github.com/openshift/cluster-kube-controller-manager-operator/blob/8be94db9da523152af2268bd7f891fe089a424eb/bindata/v4.1.0/config/defaultconfig.yaml#L8-L9) by default for all clouds, so the routes will not be provisioned for the Node by the `CCM` (which is done by `Routes()` method in the interface), and so not required for SDN functionality.
+  - Storage CSI migration will be required for OpenStack and vSphere providers
+
+Q: Should we reuse the existing cloud provider config or generate a new one?
+A: CCM config is backward compatible with the in-tree cloud provider. It means we can reuse it.
+
+Q: How to migrate PVs created by the in-tree cloud provider?
+A: [CSIMigration](https://kubernetes.io/blog/2019/12/09/kubernetes-1-17-feature-csi-migration-beta/) looks like the best option, especially if it is GA in 1.19.
```

Comment:
Volumes are migrated? What is the downgrade strategy for this step?

Reply:
1. CSI migration most probably won't GA in 1.19 upstream.
2. Storage team does not plan to support it in OCP 4.7, not even as tech preview. Therefore, if an existing cluster is upgraded to external CCM, it looses access to *all* its PVs. For freshly installed clusters, making the CSI storage class as default looks feasible in 4.7 timeframe.